feat:  docs for wordbreaker extension API (#600)
keymanapp/help.keyman.com: developer/15.0/guides/lexical-models/advanced/unicode-breaker-extension.md

        },
        breakIfMatch: false
      }],

      /*** Character class overrides for specific characters ***/
      propertyMapping: (char) => {
        let hyphens = ['\u002d', '\u2010', '\u058a', '\u30a0'];
        // treats Khmer consonants & independent vowels in the same manner
        // as the basic latin-script based alphabet
        if(char >= '\u1780' && char <= '\u17b3') {
            return "ALetter";
        } else if(hyphens.includes(char)) {
            return "Hyphen";
        } else {
          // The other Khmer characters already have useful word-breaking
          // property assignments.
          return null;
        }
      },

      /*** Declares any new, custom character classes to be recognized by the word-breaker ***/
      customProperties: ["Hyphen"]
    };

    /*** Connects all the pieces together for actual use ***/
    return wordBreakers['default'](text, customization);
  },
  // ...
};

export default source;
```

This example's customization is designed to accomplish two goals:

1. Unicode's wordbreaker does not map base Khmer consonants to any of the
  <a href="https://unicode.org/reports/tr29/#Table_Word_Break_Property_Values">

>   relevant wordbreaking character properties</a>, causing it to be treated as `"Other"`.

Author: Late-day brain just couldn't remember the Markdown for doing links.  :+1:, will fix other cases.

  The minority language in question instead wishes for Latin-script like word-breaking,
  so mapping the consonants to the same property as Latin-script consonants allows
  them to be treated similarly - in the manner they expect.

2. By default, the wordbreaker will automatically insert word boundaries before
  and after a hyphen.  To allow mid-word hyphens, we need to remap them to
  a different character property.

  While the specification itself mentions that we could
  just map hyphens to `"MidLetter"`, this example opts to define
  a custom-tailored property that ensures only hyphens are affected in order to add
  a special, end-of-context rule that may be useful when typing - the rule above

>   referencing `"eot"`.

Author: This is covered more explicitly in one of the newly-written sections.

Unicode Standard Annex #29 §4.1 Default Word Boundary Specification</a>.

Rules WB1 through WB4 of the specification will always apply first, before any
custom rules.  Custom rules will then be applied in order of their definition within
the model, with them all being applied before the specification's rules labeled WB5
and onward.

Each rule should be of the following form:

```typescript
{
  // A function that returns 'true' whenever the rule should apply
  match: (context) => {
    // ...
  },
  // Whether to prevent (false) or to enforce (true) a boundary when the rule applies.
  breakIfMatch: false
}
```

Toward this end, the `context` object received by match provides a function called
`propertyMatch` in order to define rules like those of the wordbreaker specification.
Let's take WB6 - "do not break letters across certain punctuation" - as an example.

As written in the spec, WB6 reads **AHLetter x (MidLetter | MidNumLetQ) AHLetter**.

To break that down:
- (implicit) **Any**
- **AHLetter**: `ALetter` or `Hebrew_Letter`
- **x** - "do not break"
- (either) **MidLetter** or **MidNumLetQ**: one of `MidNumLet` or `Single_Quote`
- **AHLetter**: `ALetter` or `Hebrew_Letter`

>     - These expansions are defined at https://unicode.org/reports/tr29/#WB_Rule_Macros.

Reviewer: ```suggestion

The expansions `MidNumLetQ` and `AHLetter` are defined at https://unicode.org/reports/tr29/#WB_Rule_Macros.
```

> If written as a custom rule, it takes the following form:

Reviewer: ```suggestion
If written as a custom rule, it takes the following form, using the function `context.propertyMatch`, which takes 4 parameters, matching two characters before and two characters after a potential boundary:
```

```typescript
{
  match: (context) => {

>     return context.propertyMatch(null, // no requirements:  implicit "Any".

Reviewer: ```suggestion
    return context.propertyMatch(null, // match any character:  implicit "Any".
```

                                ["ALetter", "Hebrew_Letter"],
                                // x
                                ["MidLetter", "MidNumLet", "Single_Quote"],
                                ["ALetter", "Hebrew_Letter"]);
  },
  breakIfMatch: false // do not break
}
```

Almost all of the specification's rules are of this form:  up to two characters
before a potential boundary may be considered alongside up to two characters after
the potential boundary.

#### Word-breaking property names
The names used in each array must be defined in one of the
following places:
* https://unicode.org/reports/tr29/#Table_Word_Break_Property_Values (case-insensitive)
* `customProperties` - your <a href="#define">declaration of any custom property types</a>

> * This list: `["Other", "sot", "eot"]`

Reviewer: ```suggestion
* One of the special property types `"Other"`, `"sot"`, or `"eot"`:
```

    * `Other`:  a character without an associated word-breaking property value
    * `sot`:  "start of text" - a marker indicating the beginning of the string being word-broken
    * `eot`:  "end of text" - a marker indicating the end of the string being word-broken

#### Applying the example
As an example, when determining whether or not to break the English word `don't` when
applying WB6 as written above, this is what happens near the apostrophe:

```typescript
{
  match: (context) => {

>     return context.propertyMatch(null, /* automatic match */                // "o" - ALetter

Reviewer: ```suggestion
    return context.propertyMatch(null, /* match any character */            // "o" - ALetter
```

                                ["ALetter", "Hebrew_Letter"],               // "n" - ALetter
                                // x
                                ["MidLetter", "MidNumLet", "Single_Quote"], // "'" - Single_Quote
                                ["ALetter", "Hebrew_Letter"]);              // "t" - ALetter
  },
  breakIfMatch: false // do not wordbreak at any position where the rule matches.
}
```

The rule applies at the position between the `n` and the `'` of `don't`, telling the
word-breaker not to word-break if the rule matches at that location.

```typescript
{
  match: (context) => {
    return context.propertyMatch(null, /* automatic match */                // "n" - ALetter
                                ["ALetter", "Hebrew_Letter"],               // "'" - Single_Quote
                                // x
                                ["MidLetter", "MidNumLet", "Single_Quote"], // "t" - Single_Quote
                                ["ALetter", "Hebrew_Letter"]);              //     - eot
  },
  breakIfMatch: false // do not break
}
```

The rule, however, does not apply between the `'` and the `t`, so without the spec's WB7 rule,

> a wordbreak could occur here once WB999 is reached.

Reviewer: I don't really understand what you are saying here?

Author: I've reworked that area a bit; hopefully it's clearer now?

```typescript
/*** Definition of extra word-breaking rules ***/
{
  rules: [{
    match: (context) => {
      if(context.propertyMatch(null, ["ALetter"], ["Hyphen"], ["eot"])) {
        return true;
      } else {
        /* ... */
      } else {
        return false;
      }
    },
    breakIfMatch: false
  }],

  /*** Character class overrides for specific characters ***/
  propertyMapping: (char) => {
    let hyphens = ['\u002d', '\u2010', '\u058a', '\u30a0'];
    if(hyphens.includes(char)) {
        return "Hyphen";
    } else {
      // Use the default properties for anything else.
      return null;
    }
  },

  /*** Declares any new, custom character classes to be recognized by the word-breaker ***/
  customProperties: ["Hyphen"]
}
```

Let's walk through what this simplification is trying to achieve:

1. Hyphens are mapped to their own distinct word-breaking property.
2. The custom rule prevents wordbreaking between a letter and a hyphen at the end of text.
    - It does not include any of the `MidLetter` characters.

Matching the rule against the end of text suggests that what follows the `Hyphen` character

> is the point of text insertion.  For this example, assuming that a user has just typed

Author: I'd need to add unit tests to be sure, but I believe this isn't just a suggestion - it's a practical reality!

Reference:  https://github.com/keymanapp/keyman/blob/4f4ddb61078929e9d26db2991b502d8db1073702/common/models/templates/src/tokenization.ts#L26-L34

Also of note:  this could be used as a "default extension" to achieve what this part of that function is aiming to do: https://github.com/keymanapp/keyman/blob/4f4ddb61078929e9d26db2991b502d8db1073702/common/models/templates/src/tokenization.ts#L52-L72

Author: Related unit tests: https://github.com/keymanapp/keyman/pull/7279/commits/26fb6da3abdd1df7c6a9f36ed9aaa0797468da9d

Note:  no changes were needed to the tokenization method; I was right!

`full-`, there will be no word-break on the hyphen until either more input is received or
the user changes the site of text entry.

**Important note**: you must declare any custom properties within the `customProperties` array.
If any are missing, the missing custom properties will fail to match against any word-breaking rule.

> Make sure you don't misspell it anywhere in your customization code!

Author: Revisiting this aspect and writing this note reminded me of something I've felt is "a bit off."

See https://github.com/keymanapp/keyman/pull/7279#discussion_r971493917 - I think we should be able to infer the properties, rather than require a declaration.  That said.

> Make sure you don't misspell it anywhere in your customization code!

Yeah, that part will still need to stick around.

Author: Started implementing the enhancement, then realized some limitations of the approach that posed significant issues.  I'll probably leave things as they are instead, and thus leave this note as-is.

                                        null)) {
          return true;
        } else {
          return false;
        }
      },
      breakIfMatch: false
    }
    // Similar extensions to WB9, 10, 13a, and 13b would also be needed for robustness.
    // Note: we have left "Hebrew_Letter" out of the WB5, 6, and 7 rewrites to help
    // simplify this example.
  ],
  propertyMapping: (char) => {
    const vowels = ['a', 'e', 'i', 'o', 'u'];
    // French and Italian allow accented vowels; this will strip off the accent and
    // leave us with the base vowel.
    const baseChar = char.normalize('NFD').charAt(0);
    if(vowels.includes(baseChar)) {
      return "AVowel";
    }

    return null;
  },
  customProperties: ["AVowel"]
}
```

Note that we have left out some of the rule extensions that would help cover certain
less-frequently encountered cases.  These may matter to some parts of a
language community, especially for models targeting a majority language.
Computer programmers in particular tend to care about the un-extended
rules (WB9, 10, 13a, and 13b) more than most.

Remember, remapping characters to a new word-breaking property prevents any default

> rule from handling them unless you add custom rules to re-includes them as their

Author: ```suggestion
rule from handling them unless you add custom rules to re-include them as their
```

The rule applies at the position between the `n` and the `'` of `don't`, telling the
word-breaker not to word-break if the rule matches at that location.

#### An unsuccessful rule match

> 

Reviewer: ```suggestion

Given the input `n't|` (where `|` is the caret), we can see that the rule will not match:
```

}
```

The same rule does not apply between the `'` and the `t`, so it does not apply at this position in

> the text.This is why the wordbreaking spec includes rule WB7 in addition to WB6 - some scenarios

Reviewer: ```suggestion
the text. This is why the wordbreaking spec includes rule WB7 in addition to WB6 - some scenarios
```

require multiple context property-matching attempts.

You may define any number of these rule objects in any order for use within the wordbreaker within
the `rules` array.

<aside>
<p>
<strong>Important note</strong>: When no other rule successfully matches a potential word-boundary
position, the spec's rule WB999 applies as a catch-all and will enforce a split.  Preventing a
word-break always requires a successfully-matching rule, though this is covered for most cases by
the Unicode specification's ruleset.
</p>
</aside>

Defining custom rules is a powerful tool, but it is detailed work and may be somewhat tedious
to get right.  Feel free to [ask for help at our Community Site](
https://community.software.sil.org/c/keyman/19) for assistance.

> ## <a name="map" id="map"></a> Character property remapping

Reviewer: Note that Markdown gives headings an id already, so this could be linked to with `#character-property-remapping`, making the `<a...>` unnecessary.

`\u005A` (the code for 'Z') - are assigned word-breaking property `ALetter`.  That
range covers 26 (`[26]`) characters.  (For our purposes here, the `L&` part is
irrelevant.)

As noted at the top of the file:

```
#  All code points not explicitly listed for Word_Break
#  have the value Other (XX).

# @missing: 0000..10FFFF; Other
```

### Redefining character properties

Of note from [our first example](#example):

```typescript
/*** Character class overrides for specific characters ***/
propertyMapping: (char) => {
  // ...
  // treats Khmer consonants & independent vowels in the same manner
  // as the basic latin-script based alphabet
  if(char >= '\u1780' && char <= '\u17b3') {
      return "ALetter";
  } // ...
},
```

If you search the [property definition text file](
https://www.unicode.org/Public/UCD/latest/ucd/auxiliary/WordBreakProperty.txt)
for `1780` or `17b3`, you will find neither.  These correspond to many letters
from the Khmer character set - notably, the 'base' characters used in Khmer's
grapheme clusters.  The other Khmer characters tend to attach at various positions
around these base characters.  The majority language for the script - Khmer -

> does not follow conventionalword-breaking rules; most notably, they do not add

Reviewer: ```suggestion
does not follow conventional word-breaking rules; most notably, they do not add
```
